torch/csrc/generic/Tensor.cpp

```diff
@@ -206,12 +207,13 @@ THTensor* THPTensor_(fromNumpy)(PyObject *numpy_array) {
     {
       THStoragePtr storage(THStorage_(newWithSize)(LIBRARY_STATE storage_size));
       switch (PyArray_TYPE(array)) {
-        case NPY_DOUBLE: COPY_FROM_ARRAY(double,  array, storage, storage_size); break;
-        case NPY_FLOAT:  COPY_FROM_ARRAY(float,   array, storage, storage_size); break;
-        case NPY_INT64:  COPY_FROM_ARRAY(int64_t, array, storage, storage_size); break;
-        case NPY_INT32:  COPY_FROM_ARRAY(int32_t, array, storage, storage_size); break;
-        case NPY_INT16:  COPY_FROM_ARRAY(int16_t, array, storage, storage_size); break;
-        case NPY_UINT8:  COPY_FROM_ARRAY(uint8_t, array, storage, storage_size); break;
+        case NPY_DOUBLE: COPY_FROM_ARRAY(char, array, storage, storage_size); break;
+        case NPY_FLOAT:  COPY_FROM_ARRAY(char, array, storage, storage_size); break;
+        case NPY_HALF:   COPY_FROM_ARRAY(char, array, storage, storage_size); break;
+        case NPY_INT64:  COPY_FROM_ARRAY(char, array, storage, storage_size); break;
+        case NPY_INT32:  COPY_FROM_ARRAY(char, array, storage, storage_size); break;
+        case NPY_INT16:  COPY_FROM_ARRAY(char, array, storage, storage_size); break;
+        case NPY_UINT8:  COPY_FROM_ARRAY(char, array, storage, storage_size); break;
```

Comment:
Hm, looking at this again, I made these changes with CPU in mind, but why don't they break the CUDA copy? I guess I should put in the old types again and use `float` for HALF as dummy.

Reply:
That's true, but actually I have no idea why the CUDA functions do what they do. CUDA version should also be a matter of allocating a CUDA tensor and then doing cudaMemcpy (there's no need to go through THC for this one). Can you update it? Thank you!

Reply:
Done